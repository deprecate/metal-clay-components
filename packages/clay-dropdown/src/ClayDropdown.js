import 'clay-icon';
import './ClayDropdownBase';
import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import itemsValidator from './items_validator';
import templates from './ClayDropdown.soy.js';

/**
 * Implementation for Metal Clay Dropdown.
 */
class ClayDropdown extends Component {}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayDropdown.STATE = {
	/**
	 * Button configuration to place a button at dropdown footer.
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?string}
	 * @default undefined
	 */
	button: Config.shapeOf({
		label: Config.string().required(),
		style: Config.oneOf(['primary', 'secondary']),
		type: Config.oneOf(['button', 'reset', 'submit']),
	}),

	/**
	 * Flag to indicate if menu is expanded.
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?bool}
	 * @default false
	 */
	expanded: Config.bool().value(false),

	/**
	 * Position in which item indicator symbols will be placed. Needed if any
	 * item has indicators.
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?string}
	 * @default undefined
	 */
	indicatorsPosition: Config.oneOf(['left', 'right']),

	/**
	 * List of menu items.
	 * @instance
	 * @memberof ClayDropdown
	 * @type {!Array}
	 * @default undefined
	 */
	items: itemsValidator.required(),

	/**
	 * Flag to indicate if menu has a search field and search through elements
	 * is possible.
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?bool}
	 * @default false
	 */
	searchable: Config.bool().value(false),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @instance
	 * @memberof ClayDropdown
	 * @type {!string}
	 * @default undefined
	 */
	spritemap: Config.string().required(),

	/**
	 * CSS classes to be applied to the trigger element.
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?string}
	 * @default undefined
	 */
	triggerClasses: Config.string(),

	/**
	 * Label of the trigger button.
	 * @instance
	 * @memberof ClayDropdown
	 * @type {!string}
	 * @default undefined
	 */
	label: Config.string().required(),

	/**
	 * Style of the trigger button.
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?string}
	 * @default unstyled
	 */
	style: Config.oneOf(['secondary', 'unstyled']).value('unstyled'),

	/**
	 * Type of the dropdown menu.
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?string}
	 * @default list
	 */
	type: Config.oneOf(['form', 'list']).value('list'),
};

defineWebComponent('clay-dropdown', ClayDropdown);

Soy.register(ClayDropdown, templates);

export {ClayDropdown};
export default ClayDropdown;
