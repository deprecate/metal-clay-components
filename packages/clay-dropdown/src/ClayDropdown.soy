{namespace ClayDropdown}

/**
 * This renders a dropdown.
 */
{template .render}
	{@param items: list<?>}
	{@param spritemap: string}
	{@param? button: ?}
	{@param? elementClasses: string}
	{@param? expanded: bool}
	{@param? handleButtonClick_: any}
	{@param? icon: string}
	{@param? id: string}
	{@param? itemsIconAlignment: string}
	{@param? label: html|string}
	{@param? searchable: bool}
	{@param? style: string}
	{@param? triggerClasses: string}
	{@param? type: string}

  {let $trigger kind="html"}
    {if $icon}
      {call ClayIcon.render}
        {param spritemap: $spritemap /}
        {param symbol: $icon /}
      {/call}
    {/if}

    {if $label}
      {$label}
    {/if}

    {if not $icon}
      {call ClayIcon.render}
        {param spritemap: $spritemap /}
        {param symbol: 'caret-bottom' /}
      {/call}
    {/if}
  {/let}

	{call ClayDropdownBase.render}
    {param button: $button /}
    {param elementClasses: $elementClasses /}
    {param expanded: $expanded /}
    {param events: ['buttonClicked': $handleButtonClick_] /}
    {param id: $id /}
    {param items: $items /}
    {param itemsIconAlignment: $itemsIconAlignment /}
    {param label: $trigger /}
    {param ref: 'dropdown' /}
    {param searchable: $searchable /}
    {param spritemap: $spritemap /}
    {param style: $style /}
    {param triggerClasses: $triggerClasses /}
    {param type: $type /}
  {/call}
{/template}
