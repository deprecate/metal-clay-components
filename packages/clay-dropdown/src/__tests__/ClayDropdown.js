import ClayDropdown from '../ClayDropdown';

let clayDropdown;

describe('ClayDropdown', function() {
	afterEach(() => {
		if (clayDropdown) {
			clayDropdown.dispose();
		}
	});

	it('should render a dropdown with items', () => {
		clayDropdown = new ClayDropdown({
			items: [
				{
					label: 'Item 1',
					url: 'item1url',
				},
				{
					label: 'Item 2',
					url: 'item2url',
				},
				{
					active: true,
					label: 'Item 3',
					url: 'item1ur3',
				},
				{
					disabled: true,
					label: 'Item 4',
					url: 'item1url',
				},
				{
					inputName: 'checkbox5name',
					inputValue: 'checkbox5value',
					label: 'Item 5',
					type: 'checkbox',
				},
				{
					checked: true,
					inputName: 'checkbox6name',
					inputValue: 'checkbox7value',
					label: 'Item 7',
					type: 'checkbox',
				},
				{
					disabled: true,
					inputName: 'checkbox7name',
					inputValue: 'checkbox7value',
					label: 'Item 7',
					type: 'checkbox',
				},
				{
					items: [
						{
							label: 'Item 8',
							inputValue: 'radio8value',
						},
						{
							checked: true,
							label: 'Item 9',
							inputValue: 'radio9value',
						},
						{
							disabled: true,
							label: 'Item 10',
							inputValue: 'radio10value',
						},
					],
					inputName: 'radioname',
					label: 'Radio Group',
					separator: true,
					type: 'radiogroup',
				},
			],
			label: 'Trigger',
			spritemap: 'icons.svg',
		});

		expect(clayDropdown).toMatchSnapshot();
	});

	it('should render an expanded dropdown', () => {
		clayDropdown = new ClayDropdown({
			expanded: true,
			items: [
				{
					label: 'Item 1',
					url: 'item1url',
				},
			],
			label: 'Trigger',
			spritemap: 'icons.svg',
		});

		expect(clayDropdown).toMatchSnapshot();
	});

	it('should render a dropdown with an icon trigger', () => {
		clayDropdown = new ClayDropdown({
			icon: 'list',
			items: [
				{
					label: 'Item 1',
					url: 'item1url',
				},
			],
			spritemap: 'icons.svg',
		});

		expect(clayDropdown).toMatchSnapshot();
	});

	it('should render a searchable dropdown', () => {
		clayDropdown = new ClayDropdown({
			items: [
				{
					label: 'Item 1',
					url: 'item1url',
				},
			],
			label: 'Trigger',
			searchable: true,
			spritemap: 'icons.svg',
		});

		expect(clayDropdown).toMatchSnapshot();
	});

	it('should render a dropdown with trigger custom classes', () => {
		clayDropdown = new ClayDropdown({
			items: [
				{
					label: 'Item 1',
					url: 'item1url',
				},
			],
			label: 'Trigger',
			spritemap: 'icons.svg',
			triggerClasses: 'my-custom-class',
		});

		expect(clayDropdown).toMatchSnapshot();
	});
});
