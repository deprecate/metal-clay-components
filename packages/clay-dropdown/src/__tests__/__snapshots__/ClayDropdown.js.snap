// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayDropdown should render a dropdown with items 1`] = `
<div class="dropdown">
  <button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" id="id">Trigger
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
      <use xlink:href="icons.svg#caret-bottom"></use>
    </svg>
  </button>
  <ul aria-labelledby="id" class="dropdown-menu">
    <li>
      <a class="dropdown-item" href="item1url">Item 1</a>
    </li>
    <li>
      <a class="dropdown-item" href="item2url">Item 2</a>
    </li>
    <li>
      <a class="dropdown-item active" href="item1ur3">Item 3</a>
    </li>
    <li>
      <a class="dropdown-item disabled" href="javascript:;">Item 4</a>
    </li>
    <li class="dropdown-item">
      <label class="custom-control custom-checkbox">
        <input class="custom-control-input" name="checkbox5name" value="checkbox5value" ref="input" type="checkbox">
        <span class="custom-control-indicator"></span>
        <span class="custom-control-description">Item 5</span>
      </label>
    </li>
    <li class="dropdown-item">
      <label class="custom-control custom-checkbox">
        <input checked="" class="custom-control-input" name="checkbox6name" value="checkbox7value" ref="input" type="checkbox">
        <span class="custom-control-indicator"></span>
        <span class="custom-control-description">Item 7</span>
      </label>
    </li>
    <li class="dropdown-item">
      <label class="custom-control custom-checkbox">
        <input disabled="" class="custom-control-input" name="checkbox7name" value="checkbox7value" ref="input" type="checkbox">
        <span class="custom-control-indicator"></span>
        <span class="custom-control-description">Item 7</span>
      </label>
    </li>
    <li class="dropdown-item">
      <label class="custom-control custom-radio">
        <input class="custom-control-input" name="radioname" value="radio8value" ref="input" type="radio">
        <span class="custom-control-indicator"></span>
        <span class="custom-control-description">Item 8</span>
      </label>
    </li>
    <li class="dropdown-item">
      <label class="custom-control custom-radio">
        <input checked="" class="custom-control-input" name="radioname" value="radio9value" ref="input" type="radio">
        <span class="custom-control-indicator"></span>
        <span class="custom-control-description">Item 9</span>
      </label>
    </li>
    <li class="dropdown-item">
      <label class="custom-control custom-radio">
        <input disabled="" class="custom-control-input" name="radioname" value="radio10value" ref="input" type="radio">
        <span class="custom-control-indicator"></span>
        <span class="custom-control-description">Item 10</span>
      </label>
    </li>
    <div class="dropdown-subheader">Group 11</div>
    <div class="dropdown-divider"></div>
  </ul>
</div>
`;

exports[`ClayDropdown should render a searchable dropdown 1`] = `
<div class="dropdown">
  <button aria-expanded="false" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" id="id">Trigger
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
      <use xlink:href="icons.svg#caret-bottom"></use>
    </svg>
  </button>
  <ul aria-labelledby="id" class="dropdown-menu">
    <div class="dropdown-section">
      <div class="input-group input-group-inset">
        <div class="input-group-input">
          <input class="form-control" placeholder="Search for..." type="text" ref="searchInput">
        </div>
        <span class="input-group-inset-item">
          <button class="btn btn-link">
            <svg aria-hidden="true" class="lexicon-icon lexicon-icon-search">
              <use xlink:href="icons.svg#search"></use>
            </svg>
          </button>
        </span>
      </div>
    </div>
    <li>
      <a class="dropdown-item" href="item1url">Item 1</a>
    </li>
  </ul>
</div>
`;

exports[`ClayDropdown should render an expanded dropdown 1`] = `
<div class="dropdown show">
  <button aria-expanded="true" aria-haspopup="true" class="dropdown-toggle btn btn-unstyled" id="id">Trigger
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-caret-bottom">
      <use xlink:href="icons.svg#caret-bottom"></use>
    </svg>
  </button>
  <ul aria-labelledby="id" class="dropdown-menu show" style="top: 0px; left: 0px;">
    <li>
      <a class="dropdown-item" href="item1url">Item 1</a>
    </li>
  </ul>
</div>
`;
